cookbooks/cdo-varnish/templates/default/accept-language.vcl.erb

```diff
@@ -0,0 +1,188 @@
+C{
+
+/* ------------------------------------------------------------------ */
+/* THIS FILE IS AUTOMATICALLY GENERATED BY ./gen_vcl.pl. DO NOT EDIT. */
+
+/*
+ * Accept-language header normalization
+ *
+ * - Parses client Accept-Language HTTP header
+ * - Tries to find the best match with the supported languages
+ * - Writes the best match as req.http.X-Varnish-Accept-Language
+ *
+ * First version: Cosimo, 21/Jan/2010
+ * Last update:   Cosimo, 03/Nov/2011
+ *
+ * http://github.com/cosimo/varnish-accept-language
+ *
+ */
+
+#include <ctype.h>  /* isupper */
+#include <stdio.h>
+#include <stdlib.h> /* qsort */
+#include <string.h>
+
+#define DEFAULT_LANGUAGE "en"
+#define SUPPORTED_LANGUAGES ":<%= node['cdo-apps']['i18n']['languages'].to_hash.keys.join(':') %>:"
+
+#define vcl_string char
+#define LANG_LIST_SIZE 16
+#define HDR_MAXLEN 256
+#define LANG_MAXLEN 8
+#define RETURN_LANG(x) { \
+    strncpy(lang, x, LANG_MAXLEN); \
+    return; \
+}
+#define RETURN_DEFAULT_LANG RETURN_LANG(DEFAULT_LANGUAGE)
+#define PUSH_LANG(x,y) { \
+    /* fprintf(stderr, "Pushing lang [%d] %s %.4f\n", curr_lang, x, y); */ \
+    /* We have to copy, otherwise root_lang will be the same every time */ \
+    strncpy(pl[curr_lang].lang, x, LANG_MAXLEN); \
```

Comment:
in this case, x is derived from an HTTP header so it can be arbitrarily long.

This means than pl[curr_lang].lang will lack the terminating null so calling string functions will reference off the end of the string.